examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs

```diff
@@ -0,0 +1,178 @@
+    const FOOTPRINT: u64 = T::FOOTPRINT;
+    const REQUIRES_DEEP_CLEAN_UP: bool = T::REQUIRES_DEEP_CLEAN_UP;
+
+    fn pull_spread(ptr: &mut KeyPtr) -> Self {
+        Upgradeable::new(T::pull_spread(ptr))
+    }
+
+    fn push_spread(&self, ptr: &mut KeyPtr) {
+        T::push_spread(&self.inner, ptr)
+    }
+
+    fn clear_spread(&self, ptr: &mut KeyPtr) {
+        T::clear_spread(&self.inner, ptr)
+    }
+}
+
+impl<T: PackedLayout + SpreadAllocate> SpreadLayout for Upgradeable<T, NotInitialized> {
+    const FOOTPRINT: u64 = <T as SpreadLayout>::FOOTPRINT;
+    const REQUIRES_DEEP_CLEAN_UP: bool = <T as SpreadLayout>::REQUIRES_DEEP_CLEAN_UP;
+
+    fn pull_spread(ptr: &mut KeyPtr) -> Self {
+        if ink_env::get_contract_storage::<T>(ptr.advance_by(0))
+            .expect("could not properly decode storage entry")
+            .is_none()
+        {
+            <Self as SpreadAllocate>::allocate_spread(ptr)
+        } else {
+            Upgradeable::new(<T as SpreadLayout>::pull_spread(ptr))
+        }
+    }
+
+    fn push_spread(&self, ptr: &mut KeyPtr) {
+        <T as SpreadLayout>::push_spread(&self.inner, ptr)
+    }
+
+    fn clear_spread(&self, ptr: &mut KeyPtr) {
+        <T as SpreadLayout>::clear_spread(&self.inner, ptr)
+    }
+}
+
+impl<T: PackedLayout> PackedLayout for Upgradeable<T, Initialized> {
+    fn pull_packed(&mut self, at: &Key) {
+        <T as PackedLayout>::pull_packed(&mut self.inner, at)
+    }
+
+    fn push_packed(&self, at: &Key) {
+        <T as PackedLayout>::push_packed(&self.inner, at)
+    }
+
+    fn clear_packed(&self, at: &Key) {
+        <T as PackedLayout>::clear_packed(&self.inner, at)
+    }
+}
+
+impl<T: PackedLayout + SpreadAllocate> PackedLayout for Upgradeable<T, NotInitialized> {
+    fn pull_packed(&mut self, at: &Key) {
+        <T as PackedLayout>::pull_packed(&mut self.inner, at)
+    }
+
+    fn push_packed(&self, at: &Key) {
+        <T as PackedLayout>::push_packed(&self.inner, at)
+    }
+
+    fn clear_packed(&self, at: &Key) {
+        <T as PackedLayout>::clear_packed(&self.inner, at)
+    }
+}
+
+impl<T: SpreadAllocate + PackedLayout> SpreadAllocate for Upgradeable<T, Initialized> {
+    fn allocate_spread(ptr: &mut KeyPtr) -> Self {
+        Upgradeable::new(<T as SpreadAllocate>::allocate_spread(ptr))
+    }
+}
+
+impl<T: SpreadAllocate + PackedLayout> SpreadAllocate for Upgradeable<T, NotInitialized> {
+    fn allocate_spread(ptr: &mut KeyPtr) -> Self {
+        Upgradeable::new(<T as SpreadAllocate>::allocate_spread(ptr))
+    }
+}
+
+impl<T: PackedAllocate> PackedAllocate for Upgradeable<T, Initialized> {
+    fn allocate_packed(&mut self, at: &Key) {
+        <T as PackedAllocate>::allocate_packed(&mut self.inner, at)
+    }
+}
+
+impl<T: PackedAllocate> PackedAllocate for Upgradeable<T, NotInitialized> {
+    fn allocate_packed(&mut self, at: &Key) {
+        <T as PackedAllocate>::allocate_packed(&mut self.inner, at)
+    }
+}
```

Comment:
:scream: Wow that's a lot of boilerplate code for the storage. @xgreenx Do you think this will reduce significantly with https://github.com/paritytech/ink/issues/1134 ?

Reply:
Of course, we only will use only one trait(maybe two), and that trait will be automatically implemented by `#[ink::storage_ext]`. All keys will be managed by the user during the definition of the struct, so the code should look tiny=)

Reply:
Yeah all of this boilerplate is only needed because the current system doesn't allow to change the contract root key without manually implementing `SpreadLayout`.

Reply:
The boilerplate, oh my 🙈